design/design-generic-references.md

```diff
@@ -0,0 +1,314 @@
+them as optional and pick the value from the first one found:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: multi-ref
+spec:
+  objects:
+    - apiVersion: example.com/v1
+      kind: Foo
+      name: foo
+      optional: true
+      fieldPath: spec.id
+    - apiVersion: example.com/v1
+      kind: Bar
+      name: bar
+      optional: true
+      fieldPath: spec.myId
+```
+
+The above syntax can be used as well when a value might appear in a several
+places within single object.
+
+### Value Transformation
+
+Sometimes the referable value needs to be transformed or constructed from
+several other values. `Referable` type can be enriched to support that 
+by adding `spec.mapping` block:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: service-url
+spec:
+  objects:
+    - apiVersion: example.com/v1
+      kind: Foo
+      name: foo
+      fieldPath: spec.host
+      id: host # optional variable name in template context
+    - apiVersion: example.com/v1
+      kind: Foo
+      name: foo
+      fieldPath: spec.port
+      id: port # optional variable name in template context
+  mapping:
+    value: "https://{{ .host }}:{{ .port }}"
+```
+
+The used templating engine in `mapping.value` is [Go templates]. Each object 
+reference with assigned `id` becomes available under that name in the template context.
+A number of useful transformation functions could be made available to that context.
+
+### Advantages
+
+* All referred values are declared on managed resources using already familiar
+  mechanism, keeping their discovery simple
+* Value extraction is detached from its consumption, same `Referable` can be
+  used by many multiple manged resources
+* `Referable` instances can be watched and used by non-Crossplane controllers as
+  well
+* `Referable` instances can be properly garbage collected if owner
+  or [`Usage`](https://github.com/crossplane/crossplane/blob/master/design/one-pager-generic-usage-type.md)
+  instances are declared
+* Multiple objects/fields can be referred as value sources
+* Value can be transformed using a Go template
+
+### Disadvantages
+
+* Providers might need to get read access to `Referable` type
+
+### Implementation
+
+It consists of two parts:
+
+* Adding new Crossplane controller for `Referable` type
+* Update [controller-runtime](https://github.com/crossplane/crossplane-runtime/)
+  to support usage of `Referable` instances
+
+#### Referable Controller
+
+The controller is responsible for:
+
+* Defining proper `Usage` instance to guard against improper deletion
+  of `Referable` instances
+* Retrieving the declared objects and the field value, based on the provided
+  references
+* Transforming the value (if requested) and exposing it under `status.value`
+* Marking the instance as ready
+
+We assume currently that the value does not change after it gets exposed.
```

Comment:
What would happen if the value changes?

Reply:
Allowing updates and propagating them to the managed resources is out
of the scope for this proposal version.